# Finance Calculator Tutorial

To use `financecalculator` in a project:

Welcome to the FinanceCalculator package documentation! FinanceCalculator is a Python package that can be used for calculating financial metrics specifically designed for loans or investment scenarios. This package serves as a convenient tool for managing personal finances, offering functionalities to calculate Present Value, Future Value, Contributions, and Frequency.

In [1]:
#import pandas as pd
import financecalculator
from financecalculator.future_value import future_value
from financecalculator.present_value import present_value
from financecalculator.contribution import calculate_contribution
from financecalculator.n_periods import n_periods

print(financecalculator.__version__)

0.1.0


**NOTE:** I added excess cells as a TEMPLATE to avoid merge conflicts. Please do NOT delete or add any extra cells. Please work within the 15 cells and we can tidy it up after everyone's completed their part :) 

## Meet Tommy

Tommy is a middle-aged family man juggling work, raising kids, and managing household responsibilities. Recently, he's realized he needs to take his financial planning more seriously if he wants to secure a comfortable future for himself and his family. With some loans to pay off, a growing list of goals, and a desire to save more, Tommy is looking for practical tools that can help him make smarter financial decisions.  
Enter FinanceCalculator — a simple, easy-to-use Python package designed to help people like Tommy take control of their finances. Whether he's planning for his retirement, saving for a dream family vacation, or just trying to understand how much he needs to set aside each month, FinanceCalculator is here to guide him through the process. 

## Present Value

In [2]:
present_value(1000, 5, 12, 100) #just checking function works

Present Value  Principal  Contributions  Interest Saved
0      2168.1222       1000           1200        968.1222

Tommy dreams of buying a house in the next 10 years. He currently has <span>&#36;</span>20,000 saved up, and he wants to save a total of <span>&#36;</span>100,000 to make a down payment. To achieve his goal, he plans to contribute $500 every month, and his savings account offers an annual interest rate of 4%, compounded monthly.

Using the `present_value` function, Tommy can estimate how much his savings will be worth at the end of 10 years, considering his monthly contributions.

#### Input Parameters:
1. **Principal (PV)**: 20,000 (current saving)
2. **Annual Interest Rate (r)**: 4% (provided by the bank)
3. **Total Number of Periods (n)**: 120 months (10 years * 12 months)
4. **Monthly Contribution (PMT)**: 500 (monthly saving from Tommy’s budget)    

Tommy runs the following calculation in Python:

In [3]:
import pandas as pd
from financecalculator.present_value import present_value

principal = 20000      # current saving
annual_rate = 0.04     # receive warning
n_periods = 10 * 12    # Time period (in months)
contribution = 500     # Monthly contribution

present_value(principal, annual_rate, n_periods, contribution)

C:\Users\aa224\UBC MDS\block4\524\FinanceCalculator\src\financecalculator\present_value.py:48: UserWarning: Warning: Annual rate is percentage. E.g. if you want to enter 0.05 for 5%, please enter 5.
  warnings.warn("Warning: Annual rate is percentage. E.g. if you want to enter 0.05 for 5%, please enter 5.", UserWarning)


Present Value  Principal  Contributions  Interest Saved
0   79879.163854      20000          60000    19879.163854

***Note**: `annual_rate` is calculated in percentage, so please enter the numeric part of the percentage. E.g. if you want to enter 0.05 for 5%, please enter 5.
Pandas must also be imported for the `present_value` function, as it returns back a dataframe*

Tommu get a warning! Because `annual_rate` is calculated in percentage in all the function in `FinanceCalculator` package, he need to double check the inputs. Although the function provides a result, this is not what Tom wants.

Tommy runs the following calculation in Python with right `annual_rate`:

In [4]:
import pandas as pd
from financecalculator.present_value import present_value

principal = 20000      # current saving for education
annual_rate = 4        # Annual interest rate in percentage
n_periods = 10 * 12    # Time period (in months)
contribution = 500     # Monthly contribution

present_value(principal, annual_rate, n_periods, contribution)

Present Value  Principal  Contributions  Interest Saved
0   69385.087429      20000          60000     9385.087429

Under the current plan, Tommy can reach his goal of <span>&#36;</span>69,385.09 on top of his existing principal by regularly depositing $500 over 10 years. It's still far from 100,000! Tommy should reconsider switching to a bank with a higher savings interest rate or increasing his monthly contributions.

### Future Value

Tommy wants to ensure he has enough saved up for a comfortable retirement. The `future_value` function can help him calculate how much his current savings will grow over time, based on an expected interest rate and a set number of time (in months). To get started, Tommy can fill in the function with his current savings, expected annual rate, time period, and any contributions he wants to make each month

Tommy currently has <span>&#36;</span>50,000  dollars saved up, and hopes to retire in 30 years. With an annual interest rate of 5% compounded monthly, and Tommy decides he's also going to contribute <span>&#36;</span>100  dollars each month towards his retirement.

Let’s walk through how Tommy uses the `future_value` function to plan for his retirement:

#### Input Parameters:
1. **Principal (PV)**: 50,000 (current savings)
2. **Annual Interest Rate (r)**: 5% (provided by the bank)
3. **Total Number of Periods (n)**: 360 months (30 years * 12 months)
4. **Monthly Contribution (PMT)**: 100 (monthly savings from Tommy’s budget)    

Tommy runs the following calculation in Python:

In [5]:
import pandas as pd
from financecalculator.future_value import future_value

principal = 50000     # Current savings
annual_rate = 5       # Annual interest rate in percentage
n_periods = 12*30     # time period (in months)
contribution = 100    # Monthly contribution

future_value(principal, annual_rate, n_periods, contribution)

Future Value  Principal  Contributions  Interest Earned
0     306613.08      50000          36000        220613.08

***Note**: `n_periods` is calculated in months, so to estimate number of years you must mutiply by 12 as shown above.  
Pandas must also be imported for the `future_value` function, as it returns back a dataframe*

We can see that Tommy will be able to save a total of <span>&#36;</span>306,613.08 dollars by the time of his retirement! By using the `future_value` function, Tommy can see how much his savings will grow over time, and determine he needs to contribute now to reach his retirement goal!

## Contribution

### PMT: Paying Off a Car Loan

Tommy recently purchased a car and took out a loan of $25,000. The loan has an annual interest rate of 6% (compounded monthly) and must be repaid in 5 years. Tommy wants to figure out how much he needs to pay each month to fully repay the loan by the end of the term.

#### Input Parameters:
- **Principal (PV)**: -$25,000 (a negative value because it represents a loan, not savings)
- **Future Value (FV)**: $0 (Tommy wants to pay off the loan entirely)
- **Annual Interest Rate (r)**: 6%
- **Number of Periods (n)**: 60 months (5 years * 12 months)

Tommy runs the following calculation in Python:


In [6]:
from financecalculator.contribution import calculate_contribution

principal = -25000      # Loan amount (negative value)
future_value = 0        # Goal is to pay off the loan completely
annual_rate = 6         # Annual interest rate in percentage
n_periods = 5 * 12      # Loan term (in months)

calculate_contribution(principal, future_value, annual_rate, n_periods)


483.32003823570676

To fully repay his car loan in 5 years, Tommy will need to make monthly payments of **$483.32**.

---

#### Additional Scenario: Shorter Loan Term

Tommy is curious how his monthly payment would change if he shortens the loan term to 3 years instead of 5 years. He adjusts the `n_periods` parameter accordingly:

#### Input Parameters:
- **Principal (PV)**: -$25,000
- **Future Value (FV)**: $0
- **Annual Interest Rate (r)**: 6%
- **Number of Periods (n)**: 36 months (3 years * 12 months)

Tommy runs the following calculation in Python:


In [7]:
from financecalculator.contribution import calculate_contribution

principal = -25000      # Loan amount (negative value)
future_value = 0        # Goal is to pay off the loan completely
annual_rate = 6         # Annual interest rate in percentage
n_periods = 3 * 12      # Loan term (in months)

calculate_contribution(principal, future_value, annual_rate, n_periods)


760.548436288893

If Tommy shortens the loan term to 3 years, his monthly payment will increase to **$760.55**. By comparing both scenarios, Tommy can evaluate whether paying off the loan faster is feasible within his budget.

---

### Notes:
- **Loan Representation**: The principal for a loan is negative because it represents an amount owed, not an asset.
- **Flexible Inputs**: The `calculate_contribution` function allows for adjustments in terms, rates, and goals, making it useful for diverse financial planning scenarios.


### Contribution as Negative Value

Tommy wants to withdraw $20,000 from his investment account over the next 5 years. He currently has $50,000 saved in the account, which earns an annual interest rate of 4%, compounded monthly. Tommy wants to calculate how much he can withdraw each month while ensuring his account balance reaches $30,000 by the end of 5 years.

#### Input Parameters:
- **Principal (PV)**: $50,000 (Tommy’s current savings)
- **Future Value (FV)**: $30,000 (desired account balance after withdrawals)
- **Annual Interest Rate (r)**: 4%
- **Number of Periods (n)**: 60 months (5 years * 12 months)

Tommy runs the following calculation in Python:


In [8]:
from financecalculator.contribution import calculate_contribution

principal = 50000       # Current savings
future_value = 30000    # Desired future value after withdrawals
annual_rate = 4         # Annual interest rate in percentage
n_periods = 5 * 12      # Time period (in months)

calculate_contribution(principal, future_value, annual_rate, n_periods)


-1373.321764421278

To withdraw $20,000 over the next 5 years and still maintain $30,000 in his account, Tommy can withdraw **$-1373.32** every month. The negative sign indicates that this is a withdrawal, not a deposit.

---

### Notes:
- **Contribution as Negative**: A negative contribution represents withdrawals from an account rather than deposits.
- **Real-World Use Case**: This calculation is useful for retirees or individuals planning systematic withdrawals from savings or investment accounts.


## n_periods

In [9]:
n_periods(1000, 5, 2000, 100)

TypeError: 'int' object is not callable

Persona: Tommy, a 40-year-old family man, is becoming more financially aware and is planning for his family’s future. He wants to save up for a family vacation to Hawaii in two years while maintaining his monthly expenses.

Tommy’s Financial Goal

Tommy plans to save <span>&#36;</span>2,000 over the next two years for the vacation. He already has <span>&#36;</span>1,000 dollars saved and can contribute $100 per month. His bank offers a 5% annual interest rate, compounded monthly. Tommy wants to calculate how many months it will take to reach his savings goal.

Using the n_periods function, Tommy can input his principal amount, annual interest rate, future value, and monthly contributions to determine how long he needs to save.

Using the n_periods Function

Let’s walk through how Tommy uses the function to plan his savings.

Input Parameters:
	
    1.	Principal: 1,000 dollars(current savings)
    
	2.	Annual Interest Rate: 5% (provided by the bank)
    
	3.	Future Value: 2,000 dollars(vacation goal)
    
	4.	Monthly Contribution: $100 (monthly savings from Tommy’s budget)

Tommy runs the following calculation in Python:

In [ ]:
# Tommy's input parameters
principal = 1000       # Current savings
annual_rate = 5        # Annual interest rate in percentage
future_value = 2000    # Savings goal
contribution = 100     # Monthly contribution

# Calculate the number of periods (months)
months = n_periods(principal, annual_rate, future_value, contribution)

print(f"Tommy needs to save for {months} months to reach his goal.")

This means Tommy can reach his <span>&#36;</span>2,000 goal in just 9 months if he consistently saves <span>&#36;</span>100 per month. The interest from the bank helps him achieve his goal faster.

A Reflection on Tommy’s Journey

Tommy is thrilled to realize he can afford the family vacation sooner than expected, thanks to the power of compound interest and a disciplined savings plan. The n_periods function gives him confidence in his financial planning, allowing him to focus on other goals like retirement savings and his children’s education.

How You Can Use the n_periods Function

Whether you’re saving for a vacation, a new car, or any financial goal, the n_periods function simplifies the math for you. Just input your starting savings, interest rate, goal, and monthly contributions to determine the exact time required to achieve your dreams.